### ISSUE TITLE
Problem reading unit cell data from OpenMM 8.x DCD files into CHARMM

### ISSUE THREAD

**Issue by RickVenable** [openmm/openmm#4405]
After successfully testing the performance of version 8.1 on a new H100 system, I tried some production runs and ran into trouble for analyses in CHARMM that required the unit cell information. OpenMM 7.x versions do not not present this problem, but both 8.0 and 8.1 are afflicted.

I reviewed the CHARMM code that reads DCD files and used the Linux od command to analyze the DCD files and found that while the unit cell data itself was okay, there was a setting in the initial 80 byte DCD header that was triggering the issue.  Specifically, the 20th 4-byte field is supposed to be an integer version number, which is checked to determine the storage form of the unit cell data.  Versions 21 and earlier stored the angles, instead of their cosines; OpenMM 7.x versions I checked were putting the number 24 in that field, while OpenMM 8.x versions are putting 21 in that field, which triggers the error, aborting the calculation.

I've patched a local CHARMM build to disable the version check, and the analyses can now be run without difficulty.  However, it would be appreciated if this could be corrected in the OpenMM source code.

**Comment by peastman** [comment]
It was changed in #3521 to fix the problem described in https://github.com/openmm/openmm/issues/3507#issuecomment-1065439558 and the following discussion.  It was necessary to get correct behavior with cpptraj.  Can you read through that discussion and suggest if there's a better behavior to use?

**Comment by RickVenable** [comment]
I'd like to first point out that the DCD format originated with CHARMM, and not NAMD or cpptraj.  As such, making changes that break CHARMM compatibility seem ill-considered in my view.  

NAMD and VMD in particular  have taken great liberties with the information in the file header.  It wouldn't be the first time I've made changes to the CHARMM trajectory reader to allow for the incorrect behavior of other programs, and probably won't be the last.

**Comment by peastman** [comment]
In principle I agree.  But CHARMM doesn't provide documentation on the DCD format, and the source code isn't public.  If there were a clear spec we'd follow it, but since there isn't everyone tries to reverse engineer it as best they can.

**Comment by RickVenable** [comment]
CHARMM has been freely available to academic and non-profits for quite a while, but does require registration.

If the the version number is < 22, the layout of cell  terms should be different; one should not change the version number without changing that as well.

**Comment by peastman** [comment]
Is there public documentation for the format?

**Comment by drroe** [comment]
OK - it seems like this may have been a huge case of crossed wires. @peastman does openmm store a symmetric shape matrix or the X-aligned cell parameters (X Y Z alpha beta gamma)? The behavior that @RickVenable has been seeing is apparently with OpenMM writing out the symmetric shape matrix values, but still storing the trajectory version number in the header as 21.

To be clear (and please @RickVenable correct me if I'm wrong):

- If the charmm trajectory version number is >= 22, the symmetric shape matrix should be stored (XX XY YY XZ YZ ZZ).
- If the charmm trajectory version number is < 22, the X-aligned parameters (X Y Z alpha beta gamma).

In particular the ordering of the arrays is important, so @RickVenable can you confirm the array ordering above is what charmm expects? I think (but I'm **not** 100% sure) that NAMD might write the X-aligned parameters in a different order.

**Comment by RickVenable** [comment]
Dan is correct about the versions and storage form for the unit cell data.

CHARMM and OpenMM 7.6, 7.7, 8.0, and 8.1 all appear to be using the lower third form of the shape matrix, based on my use of od to examine the DCD files.  

As noted, the order of the cell data is different for version 21 (which is about 30 years old), as well as how the angle information is stored. For the shape matrix, the angle data is stored as cos(angle) and not the angle in degrees.

There is not a definitive public listing of the DCD format that I'm aware of.  Is there a good place one could post a description of the file layout and the definitions of the fields in the header that CHARMM uses?

**Comment by peastman** [comment]
Here is the relevant code:

https://github.com/openmm/openmm/blob/ed710a0d0f0857ee2ba4de5d2157b4294d75ae87/wrappers/python/openmm/app/dcdfile.py#L160-L167

>  Is there a good place one could post a description of the file layout and the definitions of the fields in the header that CHARMM uses?

The [CHARMM documentation](https://academiccharmm.org/documentation) page would be the obvious place.  That's where I would look for it.

**Comment by RickVenable** [comment]
That looks correct, and how CHARMM will interpret the data when version.>= 22

For versions < 22, CHARMM expects to find   a_length, b_length, c_length, alpha, beta, gamma

**Comment by peastman** [comment]
What's the best behavior?  Is there a way we can write it so it will interpreted correctly in all programs?

**Comment by RickVenable** [comment]
Unclear; it seems different programs use different methods to determine the storage form of the unit cell values.  CHARMM uses the version field of the header, while others may use a heuristic test of the values, or just hope for the best.  

I don't have much experience with DCD files from programs other than CHARMM or OpenMM.

**Comment by peastman** [comment]
See the discussion following https://github.com/openmm/openmm/issues/3507#issuecomment-1065439558.  We looked at the behavior of a few different codes including NAMD, MDTraj, CPPTraj, and ProDy.  @drroe do you have any insight on it?

**Comment by peastman** [comment]
In OpenMM the box vectors are not symmetric.  Rather, three of the elements are required to be zero.  See http://docs.openmm.org/latest/userguide/theory/05_other_features.html#periodic-boundary-conditions.  Is that different from how CHARMM represents them?

**Comment by drroe** [comment]
>  Is that different from how CHARMM represents them?

So a loose description of how CHARMM generates the symmetric shape matrix is first, the the metric tensor, HtH, is calculated, where the diagonal elements are the box lengths squared, and the off diagonal elements are set to XY * cos gamma, XZ * cos beta, and YZ cos alpha (this is symmetric by definition). This is then diagonalized, and the resulting eigenvectors and eigenvalues are used to obtain the symmetric matrix. @RickVenable correct me if I'm wrong here.

**Comment by peastman** [comment]
It's probably simplest to keep writing the lengths and angles, then.  As long as that will work for all readers?

A PR fixing it would be very welcome.

**Comment by drroe** [comment]
It should work for all readers, at least VMD and CPPTRAJ (and if it works for those should work for others as well). @peastman @RickVenable does that make sense?

**Comment by RickVenable** [comment]
Version=24 would be fine, esp. as it seems other programs besides CHARMM use that as well.  VMD seemed to ignore the version field, and cleanly displayed trajectories with version=21 (it always worked for version=24).

What does OpenMM write in the unit cell record of a DCD file for a triclinic cell, _e.g._ for a truncated octahedron?

**Comment by peastman** [comment]
It writes the lengths and the sines of the angles.  See [above](https://github.com/openmm/openmm/issues/4405#issuecomment-1904868021).

**Comment by RickVenable** [comment]
That would seem to explain why a truncated octahedron looks like a box in VMD.

Displaying the triclinic form as the polyhedron it was derived from would be nice.  Is there any software that can transform back to the polyhedron and repack the cell for display purposes?

**Comment by drroe** [comment]
> Displaying the triclinic form as the polyhedron it was derived from would be nice. Is there any software that can transform back to the polyhedron and repack the cell for display purposes?

The `image familiar` command in cpptraj will do this (also `autoimage`).

**Comment by drroe** [comment]
@peastman I'm pretty sure that the way OpenMM/VMD/NAMD are storing unit cell parameters in DCD files only works with the shape matrix form if the unit cell itself is X-aligned (i.e. the Y and Z components of the A unit vector are zero). I just did some tests using a trajectory generated with CHARMM for a system with a truncated octahedron box generated by CHARMM-GUI. The truncated octahedron shape is tricky because in order for the unit cell matrix to be symmetric it needs to align a certain way (it cannot be X-aligned). If I read the charmm-generated trajectory back into charmm and ask it to print the box geometry I get:
```
0.100000      69.000000      69.000000      69.000000     109.471221     109.471221     109.471221
```
Here, the first column is time, the next 3 are lengths A B and C, and the last 3 are the angles alpha, beta, and gamma in degrees. If I take the charmm trajectory, read it into VMD (which uses the same cosine form used by NAMD and OpenMM), write it back out as a charmm DCD, then analyze that with charmm, I get the following:
```
       0.048888      66.409544      66.409544      66.409544      88.308416      88.308416      88.308416
```
Both the lengths and particularly the angles are wrong; the box has been distorted. Also note the time is wrong - I think that may be because charmm expects the timestep to be in units of 1/20.455 ps (AKMA units) but VMD is storing it in ps.

CPPTRAJ does not have this issue because it reads the full unit cell information and calculates to/from the symmetric tensor directly.

So ultimately I think OpenMM (and NAMD/VMD etc.) will need to be updated with how they store and read information to/from charmm trajectory files, at least if truncated octahedron shapes are going to work properly. But maybe this should be addressed in a new issue.

Let me know if I can provide any more information (commands, scripts, etc).

**Comment by peastman** [comment]
In OpenMM, the A vector is always x aligned.  That's a strict requirement.  It assumes that internally, and it throws an exception if you try to set unit vectors that don't satisfy it.

You can rotate the unit cell to make it point in a different direction, but if you do that you need to rotate all the particle positions as well.

**Comment by drroe** [comment]
> In OpenMM, the A vector is always x aligned. That's a strict requirement. It assumes that internally, and it throws an exception if you try to set unit vectors that don't satisfy it.

That's good and means the DCD writer should always be printing the correct values for a shape matrix. It just means that you couldn't e.g. use the initial coordinates from something like CHARMM-GUI for a truncated octahedron. Truncated octahedrons from Amber should work since those are X-aligned.

**Comment by peastman** [comment]
Fixed by #4410.
